File: app/src/main/java/com/example/softmusic/room/MusicDao.java
package com.example.softmusic.room;

import androidx.lifecycle.LiveData;
import androidx.room.Dao;
import androidx.room.Delete;
import androidx.room.Insert;
import androidx.room.Query;
import androidx.room.Transaction;
import androidx.room.Update;

import com.example.softmusic.musicSong.MusicSong;
import com.example.softmusic.songList.MusicSongList;

import java.util.List;

@Dao
public interface MusicDao {
    @Insert
    void insertMusicSong(MusicSong...songs);
    @Update
    void updateMusicSong(MusicSong...songs);
    @Delete
    void deleteMusicSong(MusicSong...songs);
    @Query("SELECT * FROM musicSong")
    LiveData<List<MusicSong>> getAllMusicSong();

    @Insert
    void insertMusicSongList(MusicSongList...songLists);
    @Update
    void updateMusicSongList(MusicSongList...songLists);
    @Delete
    void deleteMusicSongList(MusicSongList...songLists);

    @Insert
    void insertPlaylistSongCrossRef(PlaylistSongCrossRef...playlistSongCrossRefs);
    @Update
    void updatePlaylistSongCrossRef(PlaylistSongCrossRef...playlistSongCrossRefs);
    @Delete
    void deletePlaylistSongCrossRef(PlaylistSongCrossRef...playlistSongCrossRefs);


    @Query("SELECT * FROM musicSongList")
    LiveData<List<MusicSongList>> getAllMusicSongList();

    @Transaction
    @Query("SELECT * FROM musicSongList")
    public List<PlaylistWithSongs> getPlaylistsWithSongs();

    @Transaction
    @Query("SELECT * FROM musicSongList")
    public LiveData<List<PlaylistWithSongs>> getLiveDataPlaylistsWithSongs();

    @Transaction
    @Query("SELECT * FROM musicSongList WHERE songListTitle = :key")
    public LiveData<PlaylistWithSongs> getLiveDataPlayListWithKey(String key);

    @Transaction
    @Query("SELECT * FROM musicSong")
    public List<SongWithPlaylists> getSongsWithPlaylists();

    @Transaction
    @Query("SELECT * FROM musicSong")
    public LiveData<List<SongWithPlaylists>> getLiveDataSongsWithPlaylists();

}

File: app/src/main/java/com/example/softmusic/room/MusicDao.kt
package com.example.softmusic.room

import androidx.lifecycle.LiveData
import androidx.room.*
import com.example.softmusic.entity.*

@Dao
interface MusicDao {
    @Insert
    fun insertMusicSong(song: MusicSong?):Long
    @Update
    fun updateMusicSong(vararg songs: MusicSong?)

    @Delete
    fun deleteMusicSong(vararg songs: MusicSong?)

    @get:Query("SELECT * FROM musicSong")
    val allMusicSong: LiveData<List<MusicSong>>
    @Query("SELECT * FROM musicSong")
    fun getAllMusicSongs():List<MusicSong>
    @Query("SELECT mediaFileUri FROM musicsong")
    fun getAllMediaUri():List<String>
    @Query("SELECT musicSongId FROM musicsong WHERE mediaFileUri = :key")
    fun getSongIdByUri(key:String):Long

    @Query("SELECT songAlbum FROM musicsong WHERE musicSongId = :key")
    fun getImageUri(key: Long):String

    @Query("SELECT * FROM musicSong WHERE songTitle = :key")
    fun getMusicSongByKey(key:String): MusicSong
    @Query("SELECT * FROM musicSong WHERE musicSongId = :key")
    fun getMusicSongById(key:Long): MusicSong

    @Insert
    fun insertMusicSongList(vararg songLists: MusicSongList?)

    @Update
    fun updateMusicSongList(vararg songLists: MusicSongList?)

    @Delete
    fun deleteMusicSongList(vararg songLists: MusicSongList?)

    @get:Query("SELECT * FROM musicSongList")
    val allMusicSongList: LiveData<List<MusicSongList>>

    @Query("SELECT * FROM musicSongList WHERE songListTitle =:key")
    fun getMusicSongListByKey(key:String): MusicSongList
    @Query("SELECT * FROM musicSongList WHERE musicSongListId =:key")
    fun getMusicSongListById(key:Long): MusicSongList

    @Insert
    fun insertPlaylistSongCrossRef(vararg playlistSongCrossRefs: PlaylistSongCrossRef?)

    @Update
    fun updatePlaylistSongCrossRef(vararg playlistSongCrossRefs: PlaylistSongCrossRef?)

    @Delete
    fun deletePlaylistSongCrossRef(vararg playlistSongCrossRefs: PlaylistSongCrossRef?)

    @get:Query("SELECT * FROM playlistSongCrossRef")
    val allPlayListSongCrossRef:LiveData<List<PlaylistSongCrossRef>>
    @Query("SELECT * FROM playlistSongCrossRef")
    fun getAllRef():List<PlaylistSongCrossRef>


    @get:Query("SELECT * FROM musicSongList")
    @get:Transaction
    val playlistsWithSongs: List<PlaylistWithSongs?>?

    @get:Query("SELECT * FROM musicSongList")
    @get:Transaction
    val liveDataPlaylistsWithSongs: LiveData<List<PlaylistWithSongs?>?>?

    @Transaction
    @Query("SELECT * FROM musicSongList WHERE musicSongListId = :key")
    fun getLiveDataPlaylistsWithSongsById(key: Long?): LiveData<PlaylistWithSongs>

    @Transaction
    @Query("SELECT * FROM musicSongList WHERE songListTitle = :songListTitle")
    fun getPlayListsWithSongsByKey(songListTitle:String): PlaylistWithSongs

    @Transaction
    @Query("SELECT * FROM musicSongList WHERE musicSongListId = :musicSongListId")
    fun getPlayListsWithSongsById(musicSongListId:Long): PlaylistWithSongs

    @get:Query("SELECT * FROM musicSong")
    @get:Transaction
    val songsWithPlaylists: List<SongWithPlaylists?>?

    @get:Query("SELECT * FROM musicSong")
    @get:Transaction
    val liveDataSongsWithPlaylists: LiveData<List<SongWithPlaylists?>?>?
}